crates/core/src/rlp/decode.rs

```diff
@@ -8,39 +8,57 @@ use bytes::{Bytes, BytesMut};
 
 /// Trait for decoding RLP encoded slices of data.
 /// See https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp/#rlp-decoding for more information.
+/// The `decode_unfinished` method is used to decode an RLP encoded slice of data and return the decoded value along with the remaining bytes.
+/// The `decode` method is used to decode an RLP encoded slice of data and return the decoded value.
+/// Implementors need to implement the `decode_unfinished` method.
+/// While consumers can use the `decode` method to decode the RLP encoded data.
 pub trait RLPDecode: Sized {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError>;
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError>;
+
+    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+        let (decoded, remaining) = Self::decode_unfinished(rlp)?;
+        if !remaining.is_empty() {
+            return Err(RLPDecodeError::InvalidLength);
+        }
+
+        Ok(decoded)
+    }
 }
 
 impl RLPDecode for bool {
     #[inline(always)]
-    fn decode(buf: &[u8]) -> Result<Self, RLPDecodeError> {
+    fn decode_unfinished(buf: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
         let bytes = Bytes::copy_from_slice(buf);
         let len = bytes.len();
 
         if len == 0 {
```

Comment:
Can we reduce this to:
```suggestion
        if buf.is_empty() {
```